### ISSUE TITLE
在预测两个短文本相似度时，每次输入一对短文本，预测结果非常准确。但是存在一个性能问题:预测一对短文本需要花费大约5-8秒，时间太久。然后请问该如何优化代码呢？

### ISSUE THREAD

**Issue by lizhzh8** [Brokenwind/BertSimilarity#3]
**调用代码如下**
`sim.set_mode(tf.estimator.ModeKeys.PREDICT)

 predict_start_time = datetime.now()

predict = sim.predict(text_1, text_2)

predict_end_time = datetime.now()

print("预测predict花费时间：", (predict_end_time - predict_start_time).total_seconds())

score = predict[0][1]

response["score"] = str(score)
response["words"] = request_body
print('TextSimilarity time used: {} sec'.format((datetime.now() - start).total_seconds()))`

**实际结果如下：**
INFO:tensorflow:tokens: [CLS] 借 呗 逾 期 短 信 通 知 [SEP] 如 何 购 买 花 呗 短 信 通 知 [SEP]
INFO:tensorflow:input_ids: 101 955 1446 6874 3309 4764 928 6858 4761 102 1963 862 6579 743 5709 1446 4764 928 6858 4761 102 0 0 0 0 0 0 0 0 0 0 0
INFO:tensorflow:input_mask: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0
INFO:tensorflow:segment_ids: 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0
INFO:tensorflow:label: 0 (id = 0)
预测predict花费时间： 0:00:07.576646
TextSimilarity time used: 8.100247 sec

**Comment by lizhzh8** [comment]
每次调用预测接口时，都会重新 create_model()，这个花费一秒左右；`sim.predict(text_1, text_2)` 需要花费六七秒左右，那么该怎么优化呢？谢谢了

**Comment by Brokenwind** [comment]
每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？

**Comment by lizhzh8** [comment]
> 每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？

我是循环调用sim.predict(sentence1, sentence2)方法。也就是说对每一个句子对，都调用sim.predict(sentence1, sentence2)方法，等该句子对返回结果后，再把下一个句子对给sim.predict(sentence1, sentence2)方法，就是这么调用的。

看log日志显示，应该是重新 create_model()了。

**Comment by Brokenwind** [comment]
> > 每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？
> 
> 我是循环调用sim.predict(sentence1, sentence2)方法。也就是说对每一个句子对，都调用sim.predict(sentence1, sentence2)方法，等该句子对返回结果后，再把下一个句子对给sim.predict(sentence1, sentence2)方法，就是这么调用的。
> 
> 看log日志显示，应该是重新 create_model()了。

我测试一下，只是调用sim.predict(sentence1, sentence2)，并没有调用create_model，速度并不慢，1s内可以出答案。你是不是在循环内加上了 sim.set_mode(tf.estimator.ModeKeys.PREDICT) , 每调用一次set_mode就会重新create_model。

**Comment by lizhzh8** [comment]
> > > 每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？
> > 
> > 
> > 我是循环调用sim.predict(sentence1, sentence2)方法。也就是说对每一个句子对，都调用sim.predict(sentence1, sentence2)方法，等该句子对返回结果后，再把下一个句子对给sim.predict(sentence1, sentence2)方法，就是这么调用的。
> > 看log日志显示，应该是重新 create_model()了。
> 
> 我测试一下，只是调用sim.predict(sentence1, sentence2)，并没有调用create_model，速度并不慢，1s内可以出答案。你是不是在循环内加上了 sim.set_mode(tf.estimator.ModeKeys.PREDICT) , 每调用一次set_mode就会重新create_model。

嗯，循环内有 sim.set_mode(tf.estimator.ModeKeys.PREDICT)。而且 我把generate_from_queue(self)中的while true 去掉了。是这个原因吗？while true 这个我感觉是需要去掉的。
![image](https://user-images.githubusercontent.com/30767119/81662845-ba3c8d80-9470-11ea-8560-061e456ebdbe.png)

**Comment by Brokenwind** [comment]
> > > > 每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？
> > > 
> > > 
> > > 我是循环调用sim.predict(sentence1, sentence2)方法。也就是说对每一个句子对，都调用sim.predict(sentence1, sentence2)方法，等该句子对返回结果后，再把下一个句子对给sim.predict(sentence1, sentence2)方法，就是这么调用的。
> > > 看log日志显示，应该是重新 create_model()了。
> > 
> > 
> > 我测试一下，只是调用sim.predict(sentence1, sentence2)，并没有调用create_model，速度并不慢，1s内可以出答案。你是不是在循环内加上了 sim.set_mode(tf.estimator.ModeKeys.PREDICT) , 每调用一次set_mode就会重新create_model。
> 
> 嗯，循环内有 sim.set_mode(tf.estimator.ModeKeys.PREDICT)。而且 我把generate_from_queue(self)中的while true 去掉了。是这个原因吗？while true 这个我感觉是需要去掉的。
> ![image](https://user-images.githubusercontent.com/30767119/81662845-ba3c8d80-9470-11ea-8560-061e456ebdbe.png)

generate_from_queue(self)中的while true 去掉之后，预测只能跑一次，这个是不建议去掉的。

**Comment by lizhzh8** [comment]
> > > > > 每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？
> > > > 
> > > > 
> > > > 我是循环调用sim.predict(sentence1, sentence2)方法。也就是说对每一个句子对，都调用sim.predict(sentence1, sentence2)方法，等该句子对返回结果后，再把下一个句子对给sim.predict(sentence1, sentence2)方法，就是这么调用的。
> > > > 看log日志显示，应该是重新 create_model()了。
> > > 
> > > 
> > > 我测试一下，只是调用sim.predict(sentence1, sentence2)，并没有调用create_model，速度并不慢，1s内可以出答案。你是不是在循环内加上了 sim.set_mode(tf.estimator.ModeKeys.PREDICT) , 每调用一次set_mode就会重新create_model。
> > 
> > 
> > 嗯，循环内有 sim.set_mode(tf.estimator.ModeKeys.PREDICT)。而且 我把generate_from_queue(self)中的while true 去掉了。是这个原因吗？while true 这个我感觉是需要去掉的。
> > ![image](https://user-images.githubusercontent.com/30767119/81662845-ba3c8d80-9470-11ea-8560-061e456ebdbe.png)
> 
> generate_from_queue(self)中的while true 去掉之后，预测只能跑一次，这个是不建议去掉的。



> > > > > 每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？
> > > > 
> > > > 
> > > > 我是循环调用sim.predict(sentence1, sentence2)方法。也就是说对每一个句子对，都调用sim.predict(sentence1, sentence2)方法，等该句子对返回结果后，再把下一个句子对给sim.predict(sentence1, sentence2)方法，就是这么调用的。
> > > > 看log日志显示，应该是重新 create_model()了。
> > > 
> > > 
> > > 我测试一下，只是调用sim.predict(sentence1, sentence2)，并没有调用create_model，速度并不慢，1s内可以出答案。你是不是在循环内加上了 sim.set_mode(tf.estimator.ModeKeys.PREDICT) , 每调用一次set_mode就会重新create_model。
> > 
> > 
> > 嗯，循环内有 sim.set_mode(tf.estimator.ModeKeys.PREDICT)。而且 我把generate_from_queue(self)中的while true 去掉了。是这个原因吗？while true 这个我感觉是需要去掉的。
> > ![image](https://user-images.githubusercontent.com/30767119/81662845-ba3c8d80-9470-11ea-8560-061e456ebdbe.png)
> 
> generate_from_queue(self)中的while true 去掉之后，预测只能跑一次，这个是不建议去掉的。

大佬，按照你的提示，我把这两个地方改了。 一：sim.set_mode(tf.estimator.ModeKeys.PREDICT)放在外边；二：generate_from_queue(self)中的while true 加上了。
预测的速度相当的快。很给力！
![image](https://user-images.githubusercontent.com/30767119/81680634-fcbc9500-9485-11ea-8f86-1222a67680ac.png)

**Comment by lizhzh8** [comment]
> > > > > 每次调用sim.predict(sentence1, sentence2)方法时，并不会重新create_model()，你是怎么进行预测的？
> > > > 
> > > > 
> > > > 我是循环调用sim.predict(sentence1, sentence2)方法。也就是说对每一个句子对，都调用sim.predict(sentence1, sentence2)方法，等该句子对返回结果后，再把下一个句子对给sim.predict(sentence1, sentence2)方法，就是这么调用的。
> > > > 看log日志显示，应该是重新 create_model()了。
> > > 
> > > 
> > > 我测试一下，只是调用sim.predict(sentence1, sentence2)，并没有调用create_model，速度并不慢，1s内可以出答案。你是不是在循环内加上了 sim.set_mode(tf.estimator.ModeKeys.PREDICT) , 每调用一次set_mode就会重新create_model。
> > 
> > 
> > 嗯，循环内有 sim.set_mode(tf.estimator.ModeKeys.PREDICT)。而且 我把generate_from_queue(self)中的while true 去掉了。是这个原因吗？while true 这个我感觉是需要去掉的。
> > ![image](https://user-images.githubusercontent.com/30767119/81662845-ba3c8d80-9470-11ea-8560-061e456ebdbe.png)
> 
> generate_from_queue(self)中的while true 去掉之后，预测只能跑一次，这个是不建议去掉的。

不过，第一个句子对 花费时间还是挺长的。花费时间： 0:00:04.418587，用了四秒

`INFO:tensorflow:Using config: {'_model_dir': '/home/lizz/nlpapi/nlpapi/venv/nlpservice/source/bert_similarity/tmp/result/', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': gpu_options {
  per_process_gpu_memory_fraction: 0.8
  allow_growth: true
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f851af21c88>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}
start generate bert_TextSimilarity_text
WARNING:tensorflow:From /root/anaconda3/envs/dlpy36/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING:tensorflow:From /root/anaconda3/envs/dlpy36/lib/python3.6/site-packages/tensorflow/python/data/ops/dataset_ops.py:429: py_func (from tensorflow.python.ops.script_ops) is deprecated and will be removed in a future version.
Instructions for updating:
tf.py_func is deprecated in TF V2. Instead, use
    tf.py_function, which takes a python function which manipulates tf eager
    tensors instead of numpy arrays. It's easy to convert a tf eager tensor to
    an ndarray (just call tensor.numpy()) but having access to eager tensors
    means `tf.py_function`s can use accelerators such as GPUs as well as
    being differentiable using a gradient tape.
    
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:*** Features ***
INFO:tensorflow:  name = input_ids, shape = (?, 32)
INFO:tensorflow:  name = input_mask, shape = (?, 32)
INFO:tensorflow:  name = label_ids, shape = (1,)
INFO:tensorflow:  name = segment_ids, shape = (?, 32)

WARNING: The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
If you depend on functionality not listed there, please file an issue.

WARNING:tensorflow:From /home/lizz/nlpapi/nlpapi/venv/nlpservice/source/bert_similarity/modeling.py:673: dense (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dense instead.
create_model_time: 1.851312
init_checkpoint_time: 0.430665
INFO:tensorflow:**** Trainable Variables ****
EstimatorSpec_time: 0:00:00.000075
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Graph was finalized.
2020-05-12 19:14:20.368876: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2020-05-12 19:14:20.373332: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3408000000 Hz
2020-05-12 19:14:20.373813: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x7f85160dfc90 executing computations on platform Host. Devices:
2020-05-12 19:14:20.373859: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
2020-05-12 19:14:20.454853: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:998] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-05-12 19:14:20.455464: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1433] Found device 0 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7335
pciBusID: 0000:02:00.0
totalMemory: 7.93GiB freeMemory: 7.63GiB
2020-05-12 19:14:20.455491: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1512] Adding visible gpu devices: 0
2020-05-12 19:14:20.456298: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-05-12 19:14:20.456312: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990]      0 
2020-05-12 19:14:20.456319: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1003] 0:   N 
2020-05-12 19:14:20.456506: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6495 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080, pci bus id: 0000:02:00.0, compute capability: 6.1)
2020-05-12 19:14:20.457948: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x7f851689f780 executing computations on platform CUDA. Devices:
2020-05-12 19:14:20.457964: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): GeForce GTX 1080, Compute Capability 6.1
WARNING:tensorflow:From /root/anaconda3/envs/dlpy36/lib/python3.6/site-packages/tensorflow/python/training/saver.py:1266: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to check for files with this prefix.
INFO:tensorflow:Restoring parameters from /home/lizz/nlpapi/nlpapi/venv/nlpservice/source/bert_similarity/tmp/result/model.ckpt-10937
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
2020-05-12 19:14:21.954206: I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library libcublas.so.10.0 locally
INFO:tensorflow:*** Example ***
INFO:tensorflow:guid: test-0
INFO:tensorflow:tokens: [CLS] ? 怎 么 更 改 花 呗 手 机 号 码 [SEP] 我 的 花 呗 是 以 前 的 手 机 号 码 ， 怎 么 更 改 成 [SEP]
INFO:tensorflow:input_ids: 101 136 2582 720 3291 3121 5709 1446 2797 3322 1384 4772 102 2769 4638 5709 1446 3221 809 1184 4638 2797 3322 1384 4772 8024 2582 720 3291 3121 2768 102
INFO:tensorflow:input_mask: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
INFO:tensorflow:segment_ids: 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
INFO:tensorflow:label: 0 (id = 0)
花费时间： 0:00:04.418587
similarity：0.9923996329307556`
